File: cmsong111/42586.py
import math

def solution(progresses, speeds):
    count = len(speeds)
    answer = []
    result = []
    for i in range(count):
        answer.append(math.ceil((100 - progresses[i] )/ speeds[i]))
    print(answer)
    for i in range(1,count,1):
        if answer[i-1] > answer[i]:
            answer[i] = answer[i-1]
    print(answer)
    temp = {}
    for i in range(count):
        if answer[i] not in temp:
            temp[answer[i]] =1
        elif answer[i] in temp:
            temp[answer[i]] +=1
    print(temp)
    for value in temp:
        result.append(temp[value])    
    print(result)
    return result


# progresses = [93, 30, 55]
# speeds = [1, 30, 5]
# solution(progresses,speeds)
# answer = [2, 1]

# progresses = [95, 90, 99, 99, 80, 99]
# speeds = [1, 1, 1, 1, 1, 1]
# solution(progresses,speeds)
# answer = [1, 3, 2]

File: jeonghye-choi/42586.js
class Queue {
  constructor() {
    this.storage = {};
    this.front = 0;
    this.rear = 0;
  }
  size() {
    // rear가 가리키는 값이 없을 때 아무 데이터가 없는 경우
    if (this.storage[this.rear] === undefined) {
      return 0;
    } else {
      return this.rear - this.front + 1;
    }
  }

  add(value) {
    if (this.size() === 0) {
      this.storage['0'] = value;
    } else {
      this.rear += 1;
      this.storage[this.rear] = value;
    }
  }

  popFront() {
    let temp;
    if (this.front === this.rear) {
      temp = this.storage[this.front];
      delete this.storage[this.front];
      this.front = 0;
      this.rear = 0;
    } else {
      temp = this.storage[this.front];
      delete this.storage[this.front];
      this.front += 1;
    }
    return temp;
  }
}

function solution(progresses, speeds) {
  let queue = new Queue();
  progresses.forEach((progress, index) => {
    queue.add([progress, speeds[index]]);
  });

  let days = 0;
  let count;
  var answer = [];

  while (queue.size() > 0) {
    count = 0;
    while (
      queue.storage[queue.front][0] + queue.storage[queue.front][1] * days >=
      100
    ) {
      queue.popFront();
      count += 1;
      if (queue.size() === 0) {
        break;
      }
    }
    if (count !== 0) {
      answer.push(count);
    }
    days++;
  }
  return answer;
}
